### ISSUE TITLE
Change default Terminal app for "Open Shell" functionality

### ISSUE THREAD

**Issue by s-sy-y** [aptakube/aptakube#196]
### What problem are you trying to solve? Do you a suggestion on how to solve it?

Would be handy to have support for choosing which terminal app is used when using "Open Shell" functionality on containers. Instead of default Terminal on MacOS, would be great if terminals like Warp, iTerm, Hyper could be used instead.

**Comment by goenning** [comment]
hey @s-sy-y thanks for the suggestion.

Which terminal do you use? Aptakube currently supports iTerm2 and the default terminal.
As long as you have iTerm.app installed, it'll use that instead of the default.

But I agree having it configurable is better

**Comment by s-sy-y** [comment]
Hey! Currently using [Warp](https://www.warp.dev/). Didn't notice that iTerm2 is chosen as default already if installed, that's nice. iTerm is a great default, but as you agreed the ability to override would indeed be even better.  :+1:

**Comment by joseluisq** [comment]
In general, to be able to choose the terminal should be nice to have for Linux desktops as well. Since not all people use GNOME-based terminals.

BTW maybe this could help to fix https://github.com/aptakube/aptakube/issues/197?

**Comment by SiegfriedEhret** [comment]
[Kitty](https://sw.kovidgoyal.net/kitty/) user here, I would love to use it instead of the default terminal.

**Comment by nicholas-ochoa** [comment]
Another vote for `kitty` here as well. Also customizing the command generated would be helpful, the default works but I may not always want to clear the screen or echo the "Connected to container" message, for example.

**Comment by tejomaygadgil** [comment]
Option to select WSL terminal in Windows would be nice too.

Currently the "shell" button opens in PowerShell:
![image](https://github.com/aptakube/aptakube/assets/10650758/c66b6023-b79c-4c9e-bd6f-2643d478cb7c)

**Comment by SiegfriedEhret** [comment]
Alternative idea: having a button to copy the command to connect so we can paste it in our favorite terminal.

**Comment by syepes** [comment]
Effectively this would be a create addition, I prefer using https://github.com/wez/wezterm in MacOS

**Comment by syepes** [comment]
@goenning Any news on adding support or configurability to open a different terminal?

**Comment by hpoznanski** [comment]
Any news here?

**Comment by Sheffer** [comment]
KDE Debian user here. Gnome-terminal wont even run on my system without installing half (or whole) of Gnome, which is not really an option.

Can you add a popup with the command at least so we can take that and transform it to whatever works for our specific terminal?
Thanks!

**Comment by goenning** [comment]
@Sheffer im working on adding more terminals at the moment, which one you’re using?

Some Linux users have reported that if you symlink `gnome-terminal` to your actual terminal, it works. At least it worked with Kitty

**Comment by goenning** [comment]
Link to discord discussion https://discord.com/channels/1073482907496239104/1073482907496239107/1344993407747100684

**Comment by Sheffer** [comment]
Konsole, which I believe is the default in KDE. Thanks for looking into this!

**Comment by lenstr** [comment]
Support for the Ghostty terminal would be greatly appreciated!

**Comment by goenning** [comment]
Hey all, I've made some progress on supporting additional terminals.

Given the vast number of terminals out there, and the API inconsistency between them, it's very difficult for me to develop and test the integration with every terminal. 

So I've open sourced the Rust module I use to launch terminals: https://github.com/aptakube/launch-terminal-rs 
And to my surprise, it already had some contributors! 🥳

I'm also sharing here a beta release of the next version of Aptakube, which includes a few additional terminals.

**NOTE:** the official release will allow you to choose which terminal you want, but for now, Aptakube will try to find your preferred terminal by going top -> down on the list of supported terminals per OS. 

Please give it a try and let me know if you're having issues with a particular terminal.

Thank you 😊

---

macOS:
- https://releases.aptakube.com/Aptakube_1.11.0-beta2_universal.dmg

macOS Terminals:
- Ghostty
- Warp
- iTerm2
- AppleTerminal

Linux:
- https://releases.aptakube.com/aptakube_1.11.0-beta2.x86_64.rpm
- https://releases.aptakube.com/aptakube_1.11.0-beta2_amd64.AppImage
- https://releases.aptakube.com/aptakube_1.11.0-beta2_amd64.deb

Linux Terminals:
- Ghostty
- Kitty
- Warp
- GNOMETerminal
- Konsole

Windows:
- no changes

**Comment by andrewmunsell** [comment]
Looks like it's working for Warp on Mac OS (with nothing else installed), thanks!

**Comment by syepes** [comment]
Some love for MacOS [WezTerm](https://github.com/wezterm/wezterm)

**Comment by goenning** [comment]
@syepes i tried adding WezTerm, but I couldn’t find a way to do it yet.

All other terminals support this:

```
open -a warp /path/to/script.sh
```

 it’ll launch the terminal and run the script

But that doesn’t seem to work with Wez :(

If you can find a way to launch the terminal AND run a script at startup, I’d definitely add it to Aptakube too

**Comment by dustinrouillard** [comment]
Weird bug, might not even be Aptakube related, but figured I'd mention to see.

If I open Ghostty from spotlight or `/Applications` directly and then open a shell from aptakube it fails with `kubectl command not found`, if I completely quit Ghostty and then start a shell from aptakube it works fine, even with multiple shells.

I was also able to see it work if I start ghostty from another terminal or keybind using `open -a ghostty`, so seems due to a difference between starting with open and running the .app, I notice this because I start ghostty with login items, which seems to start the same way as running from spotlight since the error is present when started this way too.

**Comment by Daedaluz** [comment]
almost sounds like the path env variable is different if you launch it through the spotlight, and that all other started terminals inherit the first terminals env.

what if you start ghostty manually from spotlight, then type kubectl - does it work?

**Comment by syepes** [comment]
@goenning Looks like it can be done with https://github.com/wezterm/wezterm/issues/6759#issuecomment-2705382347

`open -a wezterm --args 'start -- /path/to/script'`

By the way, where in the GUI will we be able to select our Terminal preference?

**Comment by dustinrouillard** [comment]
> almost sounds like the path env variable is different if you launch it through the spotlight, and that all other started terminals inherit the first terminals env.
> 
> what if you start ghostty manually from spotlight, then type kubectl - does it work?

Yes, everything is normal when launched with the app, I would've noticed this haha, it only affects the terminals if they're started using aptakube after the Ghostty "daemon" is running.

**Comment by goenning** [comment]
@dustinrouillard would you mind comparing the output of `echo $PATH` between running through open (via Aptakube) and running from Spotlight?

I suppose they are different, but maybe looking at the PATH variable (and maybe others) might give us a hint at what's going on.

**Comment by dustinrouillard** [comment]
The PATH is different, however `/opt/homebrew/bin` is present and once the terminal exits to prompt after the failure I can run the command fine (Like I can literally execute `./run-in-terminal.sh` from the directory it puts me in and that works from the same window it launched)

My only guess is my PATH definition in my zshrc isn't being applied at time of execution of the initial command or something of that nature.

Hard to debug by adding an echo or something, since this shell script is overwritten on each attempt, would love to find this one.

![Image](https://github.com/user-attachments/assets/5c950aff-f9b9-4ffd-aeb1-b6b9147be0f8)

**Comment by goenning** [comment]
It’s easier to debug using this https://github.com/aptakube/launch-terminal-rs

But you’ll need to install Rust and Node to compile it.

Alternatively: modify the script and launch Ghostyy from Terminal.app using open. I believe the outcome is the same as if launching from Aptakube?

**Comment by dustinrouillard** [comment]
Good point, added a echo for $PATH before the $SHELL exec and the path results in the following.

Without ghostty already running
```
/Users/dustin/Projects/Random/launch-terminal-rs/node_modules/.bin:/Users/dustin/Projects/Random/node_modules/.bin:/Users/dustin/Projects/node_modules/.bin:/Users/dustin/node_modules/.bin:/Users/node_modules/.bin:/node_modules/.bin:/Users/dustin/.local/share/fnm/node-versions/v21.7.3/installation/lib/node_modules/npm/node_modules/@npmcli/run-script/lib/node-gyp-bin:/opt/homebrew/opt/postgresql@16/bin:/home/dustin/.local/share/pnpm:/Users/dustin/.local/state/fnm_multishells/70084_1741338633422/bin:/Users/dustin/.console-ninja/.bin:/Users/dustin/.local/bin:/Users/dustin/.node_modules_global/bin:/Users/dustin/.gem/ruby/2.6.0/bin:/opt/homebrew/opt/openjdk/bin:/opt/homebrew/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/X11/bin:/Library/Apple/usr/bin:/Users/dustin/.cargo/bin:/opt/local/bin:/Users/dustin/.bin:/Users/dustin/Library/Android/sdk/emulator:/Users/dustin/Library/Android/sdk/tools:/Users/dustin/Library/Android/sdk/tools/bin:/Users/dustin/Library/Android/sdk/platform-tools:/Users/dustin/.gopath/bin:/Applications/Ghostty.app/Contents/MacOS
```

With ghostty already opened.
```
/usr/bin:/bin:/usr/sbin:/sbin:/Applications/Ghostty.app/Contents/MacOS
```

Happens with the package, and `open` directly with the script path. Makes sense that it's PATH related and that makes it a ghostty issue, thanks for debugging help.

**Comment by elijah-rou** [comment]
@dustinrouillard you manage to resolve this? Ghostty doesn't seem to have a setting related to PATH env

**Comment by goenning** [comment]
Here's a new beta, which includes WezTerm on macOS, setting to make terminal configurable, and the option to start a new kubectl session on the selected cluster.

As always, feedback is welcome! :)

---

macOS:
- https://releases.aptakube.com/Aptakube_1.11.0-beta3_universal.dmg

Linux:
- https://releases.aptakube.com/aptakube_1.11.0-beta3.x86_64.rpm
- https://releases.aptakube.com/aptakube_1.11.0-beta3_amd64.AppImage
- https://releases.aptakube.com/aptakube_1.11.0-beta3_amd64.deb

**Comment by syepes** [comment]
@goenning With the beta3 it works but only if there is no already opened wezterm terminal, I think we would need to ass the flag `-n` so that it can open always.

`open -n -a wezterm --args start -- /Users/goenning/Desktop/test.sh`

**Comment by goenning** [comment]
@syepes can you try again with https://releases.aptakube.com/Aptakube_1.11.0-beta5_universal.dmg please?

**Comment by syepes** [comment]
Perfect, now it works without even opening an extra terminal :-)

**Comment by goenning** [comment]
This is now available on 1.11, thanks all for helping me test it ❤️
